### PR TITLE
Add `HostedZone` CRD

### PR DESCRIPTION
**Pull request by RedbackThomson**
Description of changes:
Creates the `HostedZone` CRD and adds basic smoke tests for public and private zones

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: templates/hooks/hosted_zone/sdk_create_post_build_request.go.tpl [hunk 1]
```diff
@@ -0,0 +1,6 @@
+    // You must use a unique CallerReference string every time you submit a
+    // CreateHostedZone request. CallerReference can be any unique string, for
+    // example, a date/timestamp.
+    // TODO: Name is not sufficient, since a failed request cannot be retried.
+    // We might need to import the `time` package into `sdk.go`
+    input.SetCallerReference(*desired.ko.Spec.Name)
```
Comment (RedbackThomson) [inline templates/hooks/hosted_zone/sdk_create_post_build_request.go.tpl:6]: To reviewers: Please note that this needs to be changed. Looking for suggestions

Comment (A-Hilaly) [reply templates/hooks/hosted_zone/sdk_create_post_build_request.go.tpl:6]: If you wanna avoid importing "time" package. You can use a different file like `hook.go` and create some function like `generateCallerReference`

File: pkg/resource/hosted_zone/hook.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+	svcapitypes "github.com/aws-controllers-k8s/route53-controller/apis/v1alpha1"
+)
+
+// getCallerReference will generate a CallerReference for a given hosted zone
+// using the name of the zone and the current timestamp, so that it produces a
+// unique value
+func getCallerReference(zone *svcapitypes.HostedZone) string {
+	return fmt.Sprintf("%s-%d", *zone.Spec.Name, time.Now().UnixMilli())
+}
```
Comment (jaypipes) [inline pkg/resource/hosted_zone/hook.go:15]: Not sure I am a fan of the name of this field (or having it manually set by the user vs. the SDK setting it on a session-level), but meh, this is the reasoning behind this field:

https://aws.amazon.com/builders-library/making-retries-safe-with-idempotent-APIs/

**Review by jaypipes — APPROVED** [review]
Very good, @RedbackThomson thank you!

**Review by vijtrip2 — APPROVED** [review]

**Pull request merged**